M2/Macaulay2/m2/examples.m2

```diff
@@ -28,9 +28,35 @@ capture List   := s -> capture demark_newline s
 --capture String := s -> capture' s -- output is (Boolean, String) => (Err?, Output)
 -- TODO: do this in interp.dd instead
 capture String := s -> (
-    pushvar(symbol interpreterDepth, 1);
+    -- TODO: this should eventually be unnecessary
+    oldThreadLocalVars := (gbTrace, debugLevel, errorDepth, interpreterDepth, debuggingMode, stopIfError, notify);
+
+    oldPrivateDictionary := User#"private dictionary";
+    oldDictionaryPath := dictionaryPath;
+    oldLoadedPackages := loadedPackages;
+    pushvar(symbol OutputDictionary, new GlobalDictionary);
+
+    User#"private dictionary" = new Dictionary;
+    OutputDictionary = new GlobalDictionary;
+    dictionaryPath = {
+	User#"private dictionary",
+	currentPackage.Dictionary,
+	oldPrivateDictionary,
+	Core.Dictionary,
+	OutputDictionary,
+	PackageDictionary};
+    scan(Core#"pre-installed packages", needsPackage);
+    scan(currentPackage.Options.PackageExports, needsPackage);
```

Comment:
Here we could simply do `needsPackage currentPackage`, and also remove `currentPackage.Dictionary,` from the statement above.

Reply:
The problem, I think, is that if the package is already loaded (which it always is!), the operation doesn't do anything, in particular it doesn't add the contents of PackageExports to the PackageDictionary.

Reply:
This experiment seems to show that it does, since the way it tells whether a package is loaded is by looking in `loadedPackages`:
```m2
Macaulay2, version 1.16.0.2
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems,
               LLLBases, MinimalPrimes, PrimaryDecomposition, ReesAlgebra,
               TangentCone, Truncations

i1 : save = (dictionaryPath,loadedPackages);

i2 : needsPackage "BGG"

o2 = BGG

o2 : Package

i3 : (dictionaryPath,loadedPackages) = save;

i4 : needsPackage "BGG"

o4 = BGG

o4 : Package

i5 : dictionaryPath

o5 = {BGG.Dictionary, BoijSoederberg.Dictionary, Truncations.Dictionary,
     ---------------------------------------------------------------------------------
     SimpleDoc.Dictionary, InverseSystems.Dictionary, ConwayPolynomials.Dictionary,
     ---------------------------------------------------------------------------------
     ReesAlgebra.Dictionary, TangentCone.Dictionary, Classic.Dictionary,
     ---------------------------------------------------------------------------------
     PrimaryDecomposition.Dictionary, IntegralClosure.Dictionary,
     ---------------------------------------------------------------------------------
     MinimalPrimes.Dictionary, LLLBases.Dictionary, Elimination.Dictionary,
     ---------------------------------------------------------------------------------
     User#"private dictionary", User.Dictionary, Core.Dictionary, OutputDictionary,
     ---------------------------------------------------------------------------------
     PackageDictionary}

o5 : List
```

Reply:
Oh, sure.